include/boost/gil/image.hpp

```diff
@@ -64,73 +66,163 @@ class image {
             copy_pixels(img._view,_view);
-        else {
+        else
+        {
             image tmp(img);
             swap(tmp);
         }
         return *this;
     }
 
     template <typename Img>
-    image& operator=(const Img& img) {
+    image& operator=(const Img& img)
+    {
         if (dimensions() == img.dimensions())
             copy_pixels(img._view,_view);
-        else {
+        else
+        {
             image tmp(img);
             swap(tmp);
         }
         return *this;
     }
 
-    ~image() {
+  private:
+      using propagate_allocators = std::true_type;
+      using no_propagate_allocators = std::false_type;
+
+      template <class Alloc2>
+      using choose_pocma = typename std::conditional<
+          // TODO: Use std::allocator_traits<Allocator>::is_always_equal if available
+          std::is_empty<Alloc2>::value,
+          std::true_type,
+          typename std::allocator_traits<Alloc2>::propagate_on_container_move_assignment::type
+      >::type;
+
+      static void exchange_memory(image& lhs, image& rhs)
+      {
+          lhs._memory = boost::exchange(rhs._memory, nullptr);
+          lhs._align_in_bytes = boost::exchange(rhs._align_in_bytes, 0);
+          lhs._allocated_bytes = boost::exchange(rhs._allocated_bytes, 0);
+          lhs._view = boost::exchange(rhs._view, image::view_t{});
+      };
+
+      void move_assign(image& img, propagate_allocators) noexcept {
+          // non-sticky allocator, can adopt the memory, fast
+          destruct_pixels(_view);
+          this->deallocate();
+          this->_alloc = img._alloc;
+          exchange_memory(*this, img);
+      }
+
+      void move_assign(image& img, no_propagate_allocators) {
+          if (_alloc == img._alloc) {
+              // allocator stuck to the rhs, but it's equivalent of ours, we can still adopt the memory
+              destruct_pixels(_view);
+              this->deallocate();
+              exchange_memory(*this, img);
+          } else {
+              // cannot propagate the allocator and cannot adopt the memory
+              if (img._memory)
+              {
+                  allocate_and_copy(img.dimensions(), img._view);
+                  destruct_pixels(img._view);
+                  img.deallocate();
+                  img._view = image::view_t{};
+              }
+              else
+              {
+                  destruct_pixels(this->_view);
+                  this->deallocate();
+                  this->_view = view_t{};
+              }
+          }
+      }
+    
+  public:
+      // TODO: Use noexcept(noexcept(move_assign(img, choose_pocma<allocator_type>{})))
+      // But https://gcc.gnu.org/bugzilla/show_bug.cgi?id=52869 prevents it (fixed in GCC > 9)
+      image& operator=(image&& img) {
+          if (this != std::addressof(img))
+              // Use rebinded alloc to choose pocma
+              move_assign(img, choose_pocma<allocator_type>{});
```

Comment:
What does the rebound mean here?

Reply:
It's related to:
```c++
class image {
public:
#if defined(BOOST_NO_CXX11_ALLOCATOR)
    using allocator_type = typename Alloc::template rebind<unsigned char>::other;
#else
    using allocator_type = typename std::allocator_traits<Alloc>::template rebind_alloc<unsigned char>;
#endif
```
I guess allocators could behave differently for different types (i.e. `unsigned char`), so we need to take into the rebound...

Reply:
Sorry, I got confused reading `allocator_type` as `Alloc`.

So, it is to use the rebound allocator type to control pocma selection, instead of the `Alloc` .